hail/src/main/scala/is/hail/io/fs/HadoopFS.scala

```diff
@@ -1,28 +1,94 @@
-package is.hail.utils.richUtils
+package is.hail.io.fs
 
+import is.hail.utils.{fatal, formatTime, getPartNumber, info, time, using, warn, SerializableHadoopConfiguration}
 import java.io._
+import java.util.Map
+import scala.collection.JavaConverters._
 
 import com.esotericsoftware.kryo.io.{Input, Output}
 import is.hail.io.compress.BGzipCodec
-import is.hail.utils._
+import is.hail.utils.{Context, TextInputFilterAndReplace, WithContext, readableBytes}
 import net.jpountz.lz4.{LZ4BlockOutputStream, LZ4Compressor}
 import org.apache.hadoop
-import org.apache.hadoop.fs.FileStatus
-import org.apache.hadoop.io.IOUtils._
+import org.apache.hadoop.fs.FSDataInputStream
+import org.apache.hadoop.io.IOUtils.copyBytes
 import org.apache.hadoop.io.compress.CompressionCodecFactory
 
 import scala.io.Source
 
-class RichHadoopConfiguration(val hConf: hadoop.conf.Configuration) extends AnyVal {
+class HadoopFilePath(path: hadoop.fs.Path) extends FilePath {
+  type Configuration = hadoop.conf.Configuration
 
-  def fileSystem(filename: String): hadoop.fs.FileSystem =
-    new hadoop.fs.Path(filename).getFileSystem(hConf)
+  override def toString: String = {
+    path.toString()
+  }
+
+   def getName: String = {
+    path.getName()
+  }
+
+  def getFileSystem(conf: Configuration): HadoopFileSystem = {
+    new HadoopFileSystem(path.toString, conf)
+  }
+}
+
+class HadoopFileSystem(val filename: String, conf: hadoop.conf.Configuration) extends FileSystem  {
+  private val dPath = new hadoop.fs.Path(filename)
+  private val hfs = dPath.getFileSystem(conf)
+
+  def open: FSDataInputStream = {
+    val test = hfs.open(dPath)
+    hfs.open(dPath)
+  }
+
+  def open(fPath: FilePath): FSDataInputStream = {
+    hfs.open(new hadoop.fs.Path(fPath.toString()))
+  }
+
+  def open(fPath: String): FSDataInputStream = {
+    hfs.open(new hadoop.fs.Path(fPath))
+  }
+
+  def deleteOnExit(fPath: FilePath): Boolean = {
+    hfs.deleteOnExit(new hadoop.fs.Path(fPath.toString()))
+  }
+
+  def makeQualified(fPath: FilePath): FilePath = {
+    new HadoopFilePath(hfs.makeQualified(new hadoop.fs.Path(fPath.toString())))
+  }
+
+  def makeQualified(fPath: String): FilePath = {
+    new HadoopFilePath(hfs.makeQualified(new hadoop.fs.Path(fPath.toString())))
+  }
+
+  def getPath(path: String): FilePath = {
+    new HadoopFilePath(new hadoop.fs.Path(path))
+  }
+}
 
+class HadoopFileStatus(fs: hadoop.fs.FileStatus) extends FileStatus {
```

Comment:
Need to double check that this is serializable

Reply:
Confirmed: https://hadoop.apache.org/docs/r2.7.3/api/org/apache/hadoop/io/Writable.html